### PR TITLE
Add input mocking, input sources change event

### PR DESCRIPTION
**Pull request by Manishearth**
Depends on  https://github.com/servo/webxr/pull/118

Also fixes some bugs I found.

Wanted to finish and merge this before I started on hit testing since the transient hit test stuff might have overlap.

There are a bunch of missing mock pieces that I'll probably do in a separate PR.

Still need to run tests.

Some things I skipped:
 - Doing handedness/target ray setting: See https://github.com/immersive-web/webxr-test-api/issues/46 , this would require making our impl support these changing
 - Handling button initial state: Would require some mock changes, but I ran out of time
 - Handling profiles/etc: We don't yet have impl support for these

r? @jdm

### CODE REVIEW

File: components/script/dom/xrinputsourceschangeevent.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        );
+        {
+            let event = changeevent.upcast::<Event>();
+            event.init_event(type_, bubbles, cancelable);
+        }
+
+        // let cx = global.get_cx();
+        // unsafe {
+        //     rooted!(in(*cx) let mut added_val = UndefinedValue());
+        //     added.to_jsval(*cx, added_val.handle_mut());
```
Comment (Manishearth) [inline components/script/dom/xrinputsourceschangeevent.rs]: Annoyingly, `to_jsval()` here throws a segfault at this line:

https://github.com/servo/mozjs/blob/a2957207f4b296f9ecd207d453224aafa3bd1e86/mozjs/js/src/builtin/Array.cpp#L4032

I've verified that `cx` is non null. Am I missing something?

**Comment by Manishearth** [comment]
Ready to land once servo/webxr#118 does

File: components/script/dom/fakexrdevice.rs [hunk 2]
```diff
@@ -19,21 +25,27 @@ use euclid::{RigidTransform3D, Rotation3D, Transform3D, Vector3D};
...
+    Handedness, InputId, InputSource, MockDeviceMsg, MockInputInit, MockViewInit, MockViewsInit,
+    TargetRayMode, Visibility,
+};
 
 #[dom_struct]
 pub struct FakeXRDevice {
     reflector: Reflector,
     #[ignore_malloc_size_of = "defined in ipc-channel"]
     sender: IpcSender<MockDeviceMsg>,
+    next_input_id: Cell<u32>,
```
Comment (asajeffrey) [inline components/script/dom/fakexrdevice.rs]: Nit: Make this a `Cell<InputId>`?

File: components/script/dom/webidls/XRInputSource.webidl [hunk 3]
```diff
@@ -19,7 +19,7 @@ enum XRTargetRayMode {
 [SecureContext, Exposed=Window, Pref="dom.webxr.enabled"]
 interface XRInputSource {
   readonly attribute XRHandedness handedness;
-  // [SameObject] readonly attribute XRTargetRayMode targetRayMode;
+  readonly attribute XRTargetRayMode targetRayMode;
```
Comment (asajeffrey) [inline components/script/dom/webidls/XRInputSource.webidl]: Why no `[SameObject]`?

**Review by asajeffrey — APPROVED** [review]
LGTM, just some minor nits. r=me assuming the segvs have been fixed!

File: components/script/dom/webidls/XRInputSource.webidl (hunk 3 above)
Comment (Manishearth) [reply components/script/dom/webidls/XRInputSource.webidl]: Can't SameObject enums. I believe this was fixed upstream in the spec, we just never updated the commented webidl.

**Comment by jdm** [comment]
```
./components/script/dom/webidls/FakeXRInputController.webidl:53: no newline at EOF
```

**Comment by Manishearth** [comment]
@bors-servo retry

```
3 unexpected results that are NOT known-intermittents:
  ▶ TIMEOUT [expected OK] /_webgl/conformance/glsl/misc/struct-mixed-array-declarators.html
  ▶ TIMEOUT [expected OK] /webaudio/the-audio-api/the-analysernode-interface/realtimeanalyser-basic.html
  ▶ TIMEOUT [expected OK] /css/css-align/self-alignment/parse-justify-self-003.html
```

**Comment by Manishearth** [comment]
@bors-servo retry

 - https://github.com/servo/servo/issues/24726

```
1 unexpected results that are NOT known-intermittents:
  ▶ FAIL [expected PASS] /css/CSS2/fonts/font-weight-applies-to-008.xht
  │   → /css/CSS2/fonts/font-weight-applies-to-008.xht 54a9df64f1476dd12020019d7cf22ac34d727bc0
  │   → /css/reference/pass_if_filler_text_match_bold.xht 68ca2ea769567706593ec6c195cb80a02c91b5a2
  └   → Screenshot is solid color 0xFFFFFF for /css/CSS2/fonts/font-weight-applies-to-008.xht
```

**Comment by Manishearth** [comment]
@bors-servo retry 
```
  ▶ FAIL [expected PASS] /css/CSS2/css1/c16-descendant-002.xht
  │   → /css/CSS2/css1/c16-descendant-002.xht 54a9df64f1476dd12020019d7cf22ac34d727bc0
  │   → /css/CSS2/reference/ref-this-text-should-be-green.xht cc1047e1c7ff45f6ce371b86003daf56e31e4f67
  └   → Screenshot is solid color 0xFFFFFF for /css/CSS2/css1/c16-descendant-002.xht
  ▶ FAIL [expected PASS] /_mozilla/css/transform_3d.html
  │   → /_mozilla/css/transform_3d.html 54a9df64f1476dd12020019d7cf22ac34d727bc0
  │   → /_mozilla/css/transform_3d_ref.html fc1036f05dcdf401b928df32bbf403df76bb1b9c
  └   → Screenshot is solid color 0xFFFFFF for /_mozilla/css/transform_3d.html
```

**Comment by Manishearth** [comment]
@bors-servo retry

#24726 again

**Pull request merged**